src/main/scala/viper/gobra/frontend/info/implementation/resolution/MemberResolution.scala

```diff
@@ -279,10 +279,29 @@ trait MemberResolution { this: TypeInfoImpl =>
   def tryDotLookup(b: PExpressionOrType, id: PIdnUse): Option[(Entity, Vector[MemberPath])] = {
     exprOrType(b) match {
       case Left(expr) =>
-        val methodLikeAttempt = tryMethodLikeLookup(expr, id)
-        if (methodLikeAttempt.isDefined) methodLikeAttempt
-        else tryFieldLookup(exprType(expr), id)
-
+        val (addr, nonAddr) = tryMethodLikeLookup(expr, id)
+        lazy val isGoEffAddressable = goEffAddressable(expr)
+        lazy val isEffAddressable = effAddressable(expr)
+
+        if (addr.isEmpty && nonAddr.isEmpty) {
+          // could not find the corresponding member
+          tryFieldLookup(exprType(expr), id)
+        } else if (isEffAddressable && addr.nonEmpty) {
+          addr
+        } else if (isGoEffAddressable && addr.nonEmpty && nonAddr.isEmpty) {
+          val errEntity = ErrorMsgEntity(error(id, s"$id requires a shared receiver ('share' or '@' annotations might be missing)."))
+          Some((errEntity, Vector()))
+        } else if (isEffAddressable && addr.isEmpty && nonAddr.nonEmpty) {
+          val errEntity = ErrorMsgEntity(error(id, s"$id expects a non-effectively addressable receiver, but got $expr instead"))
+          Some((errEntity, Vector()))
+        } else if (!isEffAddressable && nonAddr.nonEmpty) {
+          nonAddr
+        } else if (!isEffAddressable && addr.nonEmpty) {
+          val errEntity = ErrorMsgEntity(error(id, s"$id requires the receiver to be effectively addressable, but got $expr instead"))
+          Some((errEntity, Vector()))
+        } else {
+          Violation.violation(s"unexpected case reached: $expr")
+        }
```

Comment:
I would rearrange the cases a bit.
```suggestion
        if (addr.isEmpty && nonAddr.isEmpty) {
          // could not find the corresponding member
          tryFieldLookup(exprType(expr), id)
        } else if (isEffAddressable && addr.nonEmpty) {
          addr
        } else if (!isEffAddressable && nonAddr.nonEmpty) {
          nonAddr
        } else if (isGoEffAddressable && addr.nonEmpty && nonAddr.isEmpty) {
          val errEntity = ErrorMsgEntity(error(id, s"$id requires a shared receiver ('share' or '@' annotations might be missing)."))
          Some((errEntity, Vector()))
        } else if (!isEffAddressable && addr.nonEmpty) {
          val errEntity = ErrorMsgEntity(error(id, s"$id requires the receiver to be effectively addressable, but got $expr instead"))
          Some((errEntity, Vector()))
        } else if (isEffAddressable && addr.isEmpty && nonAddr.nonEmpty) {
          val errEntity = ErrorMsgEntity(error(id, s"$id expects a non-effectively addressable receiver, but got $expr instead"))
          Some((errEntity, Vector()))
        } else {
          Violation.violation(s"unexpected case reached: $expr")
        }
```